config/v1/types_build.go

```diff
@@ -32,17 +32,15 @@ type BuildSpec struct {
 }
 
 type BuildDefaults struct {
-	// GitHTTPProxy is the location of the HTTPProxy for Git source
+	// BuildProxy contains the proxy settings for all build operations, including image pull/push
+	// and source download.
 	// +optional
-	GitHTTPProxy string `json:"gitHTTPProxy,omitempty"`
+	BuildProxy Proxy `json:"buildProxy,omitempty"`
```

Comment:
I'd propose calling it just `Proxy`, as the general one that applies to all operations. That's my understanding. Then there will be specific ones like `GitProxy` below.

Reply:
Also can we make those pointers, since they are optional?

Reply:
My understanding of our standards is that we should use pointers if we want to distinguish between "not set" and "set to empty". I don't think we need to do that in this case, as these configs will be consumed by the openshift-cluster-openshift-controller-manger-operator and passed to the openshift controller manager's config only if they are non-empy.

Reply:
BuildProxy doesn't make sense because this whole thing is build config, plus it's unclear what aspect of "build" it applies to.  Just `Proxy` is too generic because we have gitproxy which is used when cloning source and which is separate.

So this field needs to carry a name that makes it clear it applies specifically and only to when the build is pulling+pushing images.  If we don't like `RegistryProxy` then I suggest `ImageProxy`, but I veto `Proxy` and `BuildProxy`

Reply:
Yes - it's a bit convoluted, though.
The "default" will be transformed into `*_PROXY` env vars that are added to build pod containers. See [here](https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/47/files#diff-a32feedccc2c285e54b07ce4f24afbf4R194) for WIP.

The git proxy settings will override `*_PROXY` env vars in the build pod itself. See [openshift/builder](https://github.com/openshift/builder/blob/master/pkg/build/builder/cmd/builder.go#L192).

Reply:
on the one hand i like the flexibility.  on the other hand i hate convoluted chains of hierarchy... we now have (in order of priority):

1) the gitproxy as set in the buildconfig
2) the gitproxy as set in the cluster build defaults
3) the "defaultproxy" as set in the cluster build defaults

to determine what proxy is used for git clone.  And within that you might have an "http proxy" set at one level, but a "no proxy" set at a different level, which combine to determine the overall behavior when git cloning.

I guess i like the idea of pushing a single cluster level default proxy for all build behavior enough to support "defaultproxy" though.  We just need to make sure the above precedence behavior is *very* well documented, both in the api field comments as well as the actual product doc.

Reply:
i agree with making them pointers, i don't care so much about requiring it be set explicitly at each level (at best that would be a doc'd recommendation we couldn't enforce anyway).

Reply:
switched to pointers, and updated the api docs.